### ISSUE TITLE
Roberta's Performance on BEA19-dev set and CONLL14-test set

### ISSUE THREAD

**Issue by MichaelCaohn** [grammarly/gector#45]
Hi,

I have reproduced Roberta's result on BEA19-dev set and CONLL14-test set. I discovered that for the BEA19-dev set, it's performance is higher than the result presented in the paper (about 4%-5% higher). However, for CONLL14-test set, its performance is lower than paper's result (about 3%-4% lower). 
![image](https://user-images.githubusercontent.com/22003919/89384044-15a8af80-d730-11ea-90c0-92b5b2793d35.png)
I have used the following scripts:

For Stage 1: (I used two GPUs for training)

python3 train.py --train_set=PIE/pie-9m-train-tagged.txt --dev_set=PIE/pie-9m-dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-1 --cold_steps_count=2 --accumulation_size=2 --updates_per_epoch=10000 --tn_prob=0 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0

For Stage 2:(I used 1 GPU for training)

python3 train.py --train_set=BEA19/bea19-train-tagged.txt --dev_set=BEA19/bea19-dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-2 --cold_steps_count=2 --accumulation_size=2 --updates_per_epoch=0 --tn_prob=0 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0 --pretrain_folder=model_weights/roberta-2-gpu/stage-1 --pretrain=best

For Stage3:(I used 1 GPU for training)

"python3 train.py --train_set=Stage-3-data/train-tagged.txt --dev_set=Stage-3-data/dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-3 --cold_steps_count=0 --accumulation_size=2 --updates_per_epoch=0 --tn_prob=1 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0 --pretrain_folder=model_weights/roberta-2-gpu/stage-2 --pretrain=best"

In Stage 2, the model trained for 12 epochs before it's dumped, and in Stage 3, it runs for 6 epochs. 
May I know what's the problem for this?

**Comment by komelianchuk** [comment]
Hi! Thank you for such a detailed report!
It seems that the biggest difference is happening during the stage2. (you model got a much better recall, but sacrifices precision). I think it's because the training was going somewhy too long. 
In our experiments with Roberta during stage2, it runs for 6 epochs and the best epoch was 4th.
Can you try to evaluate other stage2 models (which were dumped after 4-5 epoch)? 
BTW, are you using an Errant scorer for bea-dev and m2 scorer for CoNLL-2014?

```
  "best_epoch": 4,
  "peak_cpu_memory_MB": 8675.268,
  "peak_gpu_0_memory_MB": 3794,
  "training_duration": "3:10:05.341073",
  "training_start_epoch": 0,
  "training_epochs": 6,
  "epoch": 6,
  "training_accuracy": 0.8563338562527493,
  "training_loss": 1.0248791451851567,
  "training_cpu_memory_MB": 8675.268,
  "training_gpu_0_memory_MB": 3794,
  "validation_accuracy": 0.8524221838069531,
  "validation_loss": 1.117441440696147,
  "best_validation_accuracy": 0.8515578402531364,
  "best_validation_loss": 1.1137556071898238
```

May I also ask you to clarify which data have you used during stage2 and stage3?
What were the data sizes of train/dev?

**Comment by MichaelCaohn** [comment]
Hi, Thank you very much for the reply.
> BTW, are you using an Errant scorer for bea-dev and m2 scorer for CoNLL-2014?
For this one, yes, I am using the Errant for bea-dev and m2scorer for CONLL-2014

> May I also ask you to clarify which data have you used during stage2 and stage3?
For stage 2, I am using only the error sentences in the "fce.dev.gold.bea19.m2", "fce.test.gold.bea19.m2", "fce.train.gold.bea19.m2", "lang8.train.auto.bea19.m2", "nucle.train.gold.bea19.m2", "ABC.train.gold.bea19.m2"

For stage 3, I used all the sentences from "ABC.train.gold.bea19.m2"

>Can you try to evaluate other stage2 models (which were dumped after 4-5 epoch)?
I have tried the performance on CONLL-2014 test set, the following are the results.
model at epoch 4:
![image](https://user-images.githubusercontent.com/22003919/89404639-e9e7f280-d74c-11ea-8615-8fb13f277a62.png)
model at epoch 5:
![image](https://user-images.githubusercontent.com/22003919/89404633-e7859880-d74c-11ea-8f1d-8f638e3c8ab3.png)
model at epoch 6:
![image](https://user-images.githubusercontent.com/22003919/89404657-ee141000-d74c-11ea-816e-a648031a3d5f.png)

> What were the data sizes of train/dev?
For stage 2, the size of the training data is 1138814, dev data is 23241

**Comment by MichaelCaohn** [comment]
Hi, Thank you very much for the reply.

> BTW, are you using an Errant scorer for bea-dev and m2 scorer for CoNLL-2014?

For this one, yes, I am using the Errant for bea-dev and m2scorer for CONLL-2014

> May I also ask you to clarify which data have you used during stage2 and stage3?

For stage 2, I am using only the error sentences in the "fce.dev.gold.bea19.m2", "fce.test.gold.bea19.m2", "fce.train.gold.bea19.m2", "lang8.train.auto.bea19.m2", "nucle.train.gold.bea19.m2", "ABC.train.gold.bea19.m2"

For stage 3, I used all the sentences from "ABC.train.gold.bea19.m2"

> Can you try to evaluate other stage2 models (which were dumped after 4-5 epoch)?

I have tried the performance on CONLL-2014 test set, the following are the results.
model at epoch 4:
![image](https://user-images.githubusercontent.com/22003919/89404817-2b789d80-d74d-11ea-82c8-53ace054425e.png)
model at epoch 5:
![image](https://user-images.githubusercontent.com/22003919/89404833-33384200-d74d-11ea-90d1-9f248b9799d6.png)
model at epoch 6:
![image](https://user-images.githubusercontent.com/22003919/89404850-39c6b980-d74d-11ea-9e64-ad476419b406.png)


> What were the data sizes of train/dev?

For stage 2, the size of the training data is 1138814, dev data is 23241

**Comment by komelianchuk** [comment]
I see.
It seems like a model at epoch 5 is much closer to our results and I suggest using this checkpoint as initialization for stage3.

I also noticed that your scores after stage1 is consistently lower than ours.
How long have you been training model for the stage1?
In our case, the best epoch was the 18th.
```
{
  "best_epoch": 18,
  "peak_cpu_memory_MB": 32058.484,
  "peak_gpu_0_memory_MB": 4504,
  "training_duration": "3 days, 8:42:00.569309",
  "training_start_epoch": 0,
  "training_epochs": 19,
  "epoch": 19,
  "training_accuracy": 0.9733935128294167,
  "training_loss": 0.24086729783527552,
  "training_cpu_memory_MB": 32058.484,
  "training_gpu_0_memory_MB": 4504,
  "validation_accuracy": 0.9736692792629905,
  "validation_loss": 0.25479723660461606,
  "best_validation_accuracy": 0.973613551580457,
  "best_validation_loss": 0.2528587454151362
}
```

I've also realized that the number of epochs is a crucial parameter for all stages.
We used the early stopping mechanism (patience=3), but it seems that it is not very robust (it depends on initialization, random order of the data, and so on).
The longer you train - the better recall you get, but precision is getting worse too.
I will add the exact number of epochs that we used for all stages disregarding the early stopping parameter to the documentation.

**Comment by MichaelCaohn** [comment]
Hi,

Thank you very much for the reply. I will try using the epoch 5 for initialization for stage 3. 

> How long have you been training model for the stage1?

Stage-1 have spent me for 64hours 23minutes and 32 seconds, using two GPUs

**Comment by MichaelCaohn** [comment]
Hi, 
I just checked the training stats for Stage-1:

Below is the information
{
  "best_epoch": 19,
  "best_validation_accuracy": 0.9711143525924762,
  "best_validation_loss": 0.28540851849541066,
  "peak_cpu_memory_MB": 62156.876,
  "peak_gpu_0_memory_MB": 4090,
  "peak_gpu_1_memory_MB": 2248,
  "training_duration": "15:40:17.486700",
  "training_start_epoch": 15,
  "training_epochs": 4,
  "epoch": 19,
  "training_accuracy": 0.9723526668274357,
  "training_loss": 0.263157100161165,
  "training_cpu_memory_MB": 62156.876,
  "training_gpu_0_memory_MB": 4090,
  "training_gpu_1_memory_MB": 2248,
  "validation_accuracy": 0.9711143525924762,
  "validation_loss": 0.28540851849541066
}

**Comment by komelianchuk** [comment]
I think there is a problem with training on multiple GPUs. (if your GPU allow - consider trying to use batch_size=128 and accumulation size=2)
I would recommend you to train on single GPU as well or try to continue the training of your stage1 model for longer time (it seems that model is slightly undertrained after stage1)

**Comment by MichaelCaohn** [comment]
Thank you very much, I will try both methods :)

**Comment by MichaelCaohn** [comment]
Btw, may I know why there is some problem with training on multiple GPUs?

**Comment by skurzhanskyi** [comment]
There's no problem with multiple GPUs. It just requires changing some hyperparameters to reproduce single GPU results.

**Comment by komelianchuk** [comment]
I can not provide the exact reason, it's only my observations.
Training with batch size=128, accumulation size=2 leads in average to better results than for example batch size=32, accumulation size=8 (despite the fact that you are doing an update based on the same amount of sentences). 
I suspect it somehow related to the fact that you might have different sequence lengths in your batches, but it's only my speculations.

**Comment by MichaelCaohn** [comment]
Thank you very much for your answer

**Comment by Chhokra** [comment]
> Hi,
> 
> I have reproduced Roberta's result on BEA19-dev set and CONLL14-test set. I discovered that for the BEA19-dev set, it's performance is higher than the result presented in the paper (about 4%-5% higher). However, for CONLL14-test set, its performance is lower than paper's result (about 3%-4% lower).
> ![image](https://user-images.githubusercontent.com/22003919/89384044-15a8af80-d730-11ea-90c0-92b5b2793d35.png)
> I have used the following scripts:
> 
> For Stage 1: (I used two GPUs for training)
> 
> python3 train.py --train_set=PIE/pie-9m-train-tagged.txt --dev_set=PIE/pie-9m-dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-1 --cold_steps_count=2 --accumulation_size=2 --updates_per_epoch=10000 --tn_prob=0 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0
> 
> For Stage 2:(I used 1 GPU for training)
> 
> python3 train.py --train_set=BEA19/bea19-train-tagged.txt --dev_set=BEA19/bea19-dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-2 --cold_steps_count=2 --accumulation_size=2 --updates_per_epoch=0 --tn_prob=0 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0 --pretrain_folder=model_weights/roberta-2-gpu/stage-1 --pretrain=best
> 
> For Stage3:(I used 1 GPU for training)
> 
> "python3 train.py --train_set=Stage-3-data/train-tagged.txt --dev_set=Stage-3-data/dev-tagged.txt --model_dir=model_weights/roberta-2-gpu/stage-3 --cold_steps_count=0 --accumulation_size=2 --updates_per_epoch=0 --tn_prob=1 --tp_prob=1 --transformer_model=roberta --special_tokens_fix=1 --tune_bert=1 --skip_correct=1 --skip_complex=0 --n_epoch=20 --patience=3 --max_len=50 --batch_size=64 --tag_strategy=keep_one --cold_steps_count=0 --cold_lr=1e-3 --lr=1e-5 --predictor_dropout=0.0 --lowercase_tokens=0 --pieces_per_token=5 --vocab_path=data/output_vocabulary --label_smoothing=0.0 --pretrain_folder=model_weights/roberta-2-gpu/stage-2 --pretrain=best"
> 
> In Stage 2, the model trained for 12 epochs before it's dumped, and in Stage 3, it runs for 6 epochs.
> May I know what's the problem for this?

Any reason why "vocab_path" has been mentioned here as a parameter even in stage 1? Because, chronologically speaking before pretraining ideally there won't be any vocabulary.

**Comment by komelianchuk** [comment]
@Chhokra 
Chronologically speaking, we started experimenting with gector approach when we do not have any synthetic data. So our vocabulary is calculated on data from stage2 (all "real" publicly available data). I think it's better to use this data for building vocabulary than synthetic data.
